lib/provisioner/worker/pluginmanager.rb

```diff
@@ -21,151 +21,160 @@
+              Dir["#{File.dirname(jsonfile)}/*.rb"].each { |file| require file }
+              # check ancestor to determine plugin type
+              klass = Object.const_get(jsondata[automatortype]['classname'])
+              if klass.ancestors.include? Object.const_get('Automator')
+                raise "plugin \"#{p_name}\" attempting to load duplicate automator type \"#{automatortype}\"" if @automatormap.key?(automatortype)
+                @automatormap.merge!({automatortype => jsondata[automatortype]})
+              else
+                raise "Declared automator \"#{automatortype}\" implementation class \"#{jsondata[automatortype]['classname']}\" must extend Automator class"
+              end
+            end
+          rescue JSON::ParserError => e
+            log.error "Could not load plugin, invalid json at #{jsonfile}"
+            log.error e.message
+            log.error e.backtrace.inspect
+            @load_errors.push("Could not load plugin, invalid json at #{jsonfile}")
+            next
+          rescue => e
+            log.error "Could not load plugin at #{jsonfile}"
+            log.error e.message
+            log.error e.backtrace.inspect
+            @load_errors.push("Could not load plugin at #{jsonfile}")
+            next
           end
         end
+      end
+
+      def register_plugins(uri)
+        @providermap.each do |name, json_obj|
+          register_plugintype(name, json_obj, "#{uri}/v2/plugins/providertypes/#{name}")
+        end
+        @automatormap.each do |name, json_obj|
+          register_plugintype(name, json_obj, "#{uri}/v2/plugins/automatortypes/#{name}")
+        end
+      end
 
-        p_automatortypes.each do |automatortype|
-          raise "declared automatortype \"#{automatortype}\" is not defined" unless jsondata.key?(automatortype)
-          raise "declared automatortype \"#{automatortype}\" already defined in another plugin" if @providermap.key?(automatortype)
-
-          raise "automatortype \"#{automatortype}\" does not define an implentor classname" unless jsondata[automatortype].key?('classname')
-          # require every .rb file in the plugin top-level directory
-          Dir["#{File.dirname(jsonfile)}/*.rb"].each { |file| require file }
-          # check ancestor to determine plugin type
-          klass = Object.const_get(jsondata[automatortype]['classname'])
-          if klass.ancestors.include? Object.const_get('Automator')
-            raise "plugin \"#{p_name}\" attempting to load duplicate automator type \"#{automatortype}\"" if @automatormap.key?(automatortype)
-            @automatormap.merge!({automatortype => jsondata[automatortype]})
+      def register_plugintype(name, json_obj, uri)
+        begin
+          log.debug "registering provider/automator type: #{name}"
+          json = JSON.generate(json_obj)
+          resp = Coopr::RestHelper.put("#{uri}", json, :'Coopr-UserID' => "admin", :'Coopr-TenantID' => "superadmin")
```

Comment:
Assign the user/tenant somewhere and use it here, versus these hard-coded values. It'll make changing them later easier.